### ISSUE TITLE
Should bfloat16 have the same type promotion rule as half for  cuda device ？

### ISSUE THREAD

**Issue by GuWei007** [pytorch/pytorch#102726]
### 🚀 The feature, motivation and pitch

PyTorch ops have  type promotion rules, such as aten::bmm(cublas), aten::baddbmm(cublas), aten::matmul(cublas),aten::softmax,aten::sum. The type promotion rules of bf16 and half, some rules are the same, some are different, Should bfloat16 have the same type promotion rule as half for  cuda device.

Here is the PyTorch code:
```c++
// Different  type promotion rules
if (self.scalar_type() == dtype1 ||
      (self.is_cuda() && self.scalar_type() == kHalf && dtype1 == kFloat)) {
    return TensorIterator::reduce_op(viewed_result1, viewed_result2, self);
  }

static TensorIterator make_reduction(
    const char* name, Tensor& result, const Tensor& self,
    c10::optional<IntArrayRef> dim, bool keepdim, ScalarType out_dtype) {
  // special case for type promotion in mixed precision, improves computational
  // efficiency.
  // not generalize this to common mismatched input/output types to avoid cross
  // product of templated kernel launches.
  const bool gpu_f16_to_f32 = (
    self.is_cuda() && self.scalar_type() == kHalf && out_dtype == kFloat);
  auto in_dtype = gpu_f16_to_f32 ? self.scalar_type() : out_dtype;
  return make_reduction(name, result, self, dim, keepdim, in_dtype, out_dtype);
}

// Same type promotion rules
template <>
void gemm<at::Half>(CUDABLAS_GEMM_ARGTYPES(at::Half)) 
void gemm<at::BFloat16>(CUDABLAS_GEMM_ARGTYPES(at::BFloat16)) 


```

**Comment by GuWei007** [comment]
@svenstaro @infil00p @JackDanger @eklitzke

**Comment by ngimel** [comment]
You are pasting nonexsitent code from an old version, now these variables are called  gpu_lowp_to_f32 and check for bf16. Please next time link code in the master (github allows to do that) and reopen with a concrete example in the user code (not links to the code) that produce unexpected results

**Comment by GuWei007** [comment]
> You are pasting nonexsitent code from an old version, now these variables are called gpu_lowp_to_f32 and check for bf16. Please next time link code in the master (github allows to do that) and reopen with a concrete example in the user code (not links to the code) that produce unexpected results

My bad, I will check the code in the master, thanks very much
